### PR TITLE
Switch pp

### PR DESCRIPTION
**Pull request by natalieisak**
pretty printer using pp library

### CODE REVIEW

File: test/test.ml [hunk 1]
```diff
@@ -85,20 +101,35 @@ let bad_test f file () =
...
+
+let example_path l =
+  let root = Filename.concat ".." "examples" in
+  List.fold_left l ~init:root ~f:Filename.concat
 
-let () =
+let () = 
+  (* Format.printf "%s" (get_name ["../examples"] (example_path ["checker_tests"; "good"; "coree.p4"])); 
+     Format.printf "%s" "----------";
+     Format.printf "%s" ((Parse.parse_string (get_name ["../examples"] (example_path ["checker_tests"; "good"; "coree.p4"]))) |> Prettypp.format_program |> to_string);   *)
```
Comment (hackedy) [inline test/test.ml:115]: Can you delete this commented out code and get rid of coree.p4? It's a duplicate of pragma-string.p4 in the same directory anyway.

File: lib/test_pp [hunk 2]
```diff
@@ -0,0 +1,5 @@
+open Core_kernel
+open Util 
+open Pp
+open StdLabels
+open prettypp
```
Comment (hackedy) [inline lib/test_pp:5]: Is this a stray file?

File: lib/pp.ml [hunk 3]
```diff
@@ -0,0 +1,176 @@
+module List = struct
```
Comment (hackedy) [inline lib/pp.ml:1]: We should use `pp` from OPAM--try adding it to the dependencies in `petr4.opam` (the "depends" clause) and `lib/dune` (the `(libraries ...)` list) and deleting this file.

File: lib/pretty.ml [hunk 4]
```diff
@@ -11,7 +11,7 @@
  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
```
Comment (hackedy) [inline lib/pretty.ml:11]: Let's delete this old pretty printer and replace any uses of it with the new one.

File: lib/types.ml [hunk 5]
```diff
@@ -46,7 +46,7 @@ end
 type name =
```
Comment (hackedy) [inline lib/types.ml:46]: This all looks like whitespace changes, please revert

File: test/test.ml [hunk 6]
```diff
@@ -85,20 +101,35 @@ let bad_test f file () =
...
   let open Alcotest in
   run "Tests" [
-    "excluded tests good", (Stdlib.List.map (fun name ->
+    (* "excluded tests good", (Stdlib.List.map (fun name ->
         test_case name `Quick (excl_test name)) excluded_good_files);
```
Comment (hackedy) [inline test/test.ml:119]: Let's uncomment the other tests. You can run just the pp tests by running `dune exec -- ./test.exe test pp` from inside the `test/` folder. The last argument there, `pp`, can be any regular expression you want that matches "round trip pp tests good" (or whatever tests you want to run).

**Review by hackedy — CHANGES_REQUESTED** [review]
Thanks for doing this and sorry for taking so long reviewing the PR! I took a look at the diff--there's a few little things to fix before merging. I also played around with the pretty printer and tried to make a list in the rest of this comment of issues I found. It would be good to fix these before merging but I think some of them (e.g., parenthesizing arithmetic with as few parentheses as necessary) are probably more effort than they're worth.

For example here's an excerpt from `dune exec -- petr4 typecheck -I examples examples/checker_tests/good/psa-portid-using-newtype2.p4`. I wouldn't expect this to put a newline after the `&`.
```
        standard_metadata.egress_spec = (PortId_t)(((PortIdUInt_t)(standard_metadata.egress_spec)
                                                       &
                                                       (PortIdUInt_t)(mask)));
```

Another thing is that lists (of parameters or arguments) look like `a,b,c` instead of `a, b, c`.

There's funny whitespace around table keys. I would write this like `meta.data[0].da: exact;`
```
        key = {
            meta.data[0].da : exact ;
        }
```

Finally there's things we already talked about with wrapping and parenthesization. In core.p4 right now the pretty printer does this.
```
    void extract<
    T0>(out T0 variableSizeHeader,in bit<32> variableFieldSizeInBits);
```
I'd prefer something like this, which is actually less wide than what the pretty printer is doing right now.
```
    void extract<T0>(out T0 variableSizeHeader,
                     in bit<32> variableFieldSizeInBits);
```

And here's an example of over-parenthesization of arithmetic expressions. `petr4 typecheck -I examples examples/checker_tests/good/scalarmeta-bmv2.p4` includes
```
        meta.counter = (meta.counter - 1);
```

**Comment by natalieisak** [comment]
Hi @jnfoster! This version includes the most recent edits, if you'd like to take a look. I think the printer is getting there, except for mainly 3 issues. I'll keep playing around with it this week, but if you see any fixes that'd be great!  
1)  Random splits in types. Ex: 
extern void hash<O,T8,D,
M>(out O result,in HashAlgorithm algo,in T8 base,in D data,in M max);

2) Opening brackets going to new lines. Ex: 
apply
    {
    packet_io_ingress.apply(hdr, fabric_metadata, standard_metadata);
    filtering.apply(hdr, fabric_metadata, standard_metadata);
    forwarding.apply(hdr, fabric_metadata, standard_metadata);
  }

3) Breaks happening earlier than expected with parameters. Ex:
control
  PacketIoIngress(inout parsed_headers_t hdr,
                  inout fabric_metadata_t fabric_metadata,
                  inout standard_metadata_t standard_metadata)
  {

Also, @hackedy you suggested earlier today using a vbox for parameter declarations. I did some thinking and pp has an hvbox, which prints everything horizontally, unless it can't fit (then it'll print everything vertically). I think using the vbox vs hvbox is a stylistic choice - do you have a preference?

**Comment by hackedy** [comment]
Using an hvbox sounds right to me. I think I missed that in pp.mli.

**Comment by hackedy** [comment]
I'll take one more look at this today and see if I can fix some of the weird breaks in function/action/control declarations that Natalie mentioned in the meeting earlier.

**Review by hackedy — APPROVED** [review]

**Pull request merged**